File: csharp/Generator.cs
namespace WildlingLib;

public sealed class Generator
{
    private readonly List<Token> _tokens;
    private readonly int _count;

    public Generator(string inputPattern, IReadOnlyDictionary<string, List<string>>? dictionaries)
    {
        Source = inputPattern;
        _tokens = ParsePattern.Parse(inputPattern, dictionaries);
        var total = 1;
        foreach (var token in _tokens)
        {
            total *= token.Count();
        }
        _count = total;
    }

    public string Source { get; }

    public int Count() => _count;

    public IReadOnlyList<Token> Tokens() => _tokens;

    public string Get(int index)
    {
        if (index > _count - 1 || index < 0)
        {
            return "";
        }

        var parts = new string[_tokens.Count];
        var indexWithOffset = index;
        for (var i = 0; i < _tokens.Count; i++)
        {
            var token = _tokens[i];
            parts[i] = token.Get(indexWithOffset % token.Count());
            indexWithOffset /= token.Count();
        }
        return string.Concat(parts);
    }
}

File: fsharp/Generator.fs
namespace WildlingLib

open System.Collections.Generic

type Generator(inputPattern: string, dictionaries: IReadOnlyDictionary<string, ResizeArray<string>> option) =
    let tokens = ParsePattern.parse inputPattern dictionaries

    let count =
        let mutable total = 1
        for token in tokens do
            total <- total * token.Count()
        total

    member _.Source = inputPattern
    member _.Count() = count
    member _.Tokens() = tokens :> IReadOnlyList<_>

    member _.Get(index: int) =
        if index > count - 1 || index < 0 then
            ""
        else
            let parts = Array.zeroCreate<string> tokens.Count
            let mutable indexWithOffset = index
            for i = 0 to tokens.Count - 1 do
                let token = tokens.[i]
                parts.[i] <- token.Get(indexWithOffset % token.Count())
                indexWithOffset <- indexWithOffset / token.Count()
            String.concat "" parts
